### ISSUE TITLE
compiletest: changes in compiletest does not trigger test rerun

### ISSUE THREAD

**Issue by vakaras** [rust-lang/rust#56280]
If I modify a file in the rustc source code (`src/librustc_c*`), `./x.py test` is smart enough to rerun all tests. However, if I modify something in `src/tools/compiletest`, then `./x.py test` still ignores the tests that completed successfully in the previous run.

Steps to reproduce:
1. Run tests:
```bash
git clone git@github.com:rust-lang/rust.git
./x.py test
```
2. Modify `compiletest`. For example, replace the `expected_output_path` function in [`runtest.rs`](https://github.com/rust-lang/rust/blob/451987d86c89b38ddd8c4c124f1b9b6d4ded6983/src/tools/compiletest/src/runtest.rs#L3072) with:
```rust
    fn expected_output_path(&self, _kind: &str) -> PathBuf {
        panic!();
    }
```
3. Run tests again:
```bash
./x.py test
```

Expected output: some tests fail.

Actual output:
```plain
Build completed successfully in 0:13:13
```

**Comment by vakaras** [comment]
I would be interested in fixing this issue myself if someone could give me some guidance on how to do that. The main thing that I do not understand is how `./x.py test` understands that `rustc` was changed. It seems that it does not use `build/x86_64-unknown-linux-gnu/test/ui/*/stamp` files for this because they include only the hash of the configuration options which do not include `rustc` version.

cc @Mark-Simulacrum

**Comment by nagisa** [comment]
The most likely culprit is `fn up_to_date` in `compiletest::main` module. It is called from `compiletest::main::make_test` and will set the `ignore` flag for the test depending on its output.

**Comment by nagisa** [comment]
That being said, I find it hard to see how to adjust compiletest itself to take the dependencies into account. rustbuild should probably have some sort of dependency edge and delete the stamp files if/when stuff like compiletest changes.

**Comment by vakaras** [comment]
Compiletest source code has only 9 files. Would it be acceptable to check that their timestamps are newer than the timestamp of the stamp file?

PR: https://github.com/rust-lang/rust/pull/56680/files
